src/inmanta/env.py

```diff
@@ -1237,11 +1237,47 @@ def install_from_list(
         upgrade: bool = False,
         upgrade_strategy: PipUpgradeStrategy = PipUpgradeStrategy.ONLY_IF_NEEDED,
     ) -> None:
-        if not self.__using_venv:
+        if not self._using_venv:
             raise Exception(f"Not using venv {self.env_path}. use_virtual_env() should be called first.")
         super(VirtualEnv, self).install_from_list(requirements_list, upgrade=upgrade, upgrade_strategy=upgrade_strategy)
 
 
+class ReentrantVirtualEnv(VirtualEnv):
+    """
+    A virtual env that can be de-activated and re-activated
+
+    This allows faster reloading due to improved caching of the working set
+
+    This is intended for use in testcases to require a lot of venv switching
+    """
+
+    def __init__(self, env_path: str) -> None:
+        super(ReentrantVirtualEnv, self).__init__(env_path)
+        self.working_set = None
+
+    def deactivate(self):
+        self._using_venv = False
+        self.working_set = pkg_resources.working_set
+
+    def use_virtual_env(self) -> None:
+        """
+        Activate the virtual environment.
+        """
+        if self._using_venv:
+            # We are in use, just ignore double activation
+            return
+
+        if not self.working_set:
+            # First run
+            super().use_virtual_env()
+        else:
+            # Later run
+            self._activate_that()
+            mock_process_env(python_path=self.python_path)
+            pkg_resources.working_set = self.working_set
+            self._using_venv = True
```

Comment:
notify chance is fairly cheap, but it destroys the working set, which contains the cache we want to preserve.

Reply:
It actually destroys it, or does it just drop the reference? If it's the latter (which is what I assumed up till now) we can just reassign it, can't we?

Reply:
It drops the reference, so we could rebuild it prior to replace, but it will be less efficient (it is only relatively cheap and we do it 900+ times),